File: python/p5.py
import pdb

def get_col(s):
    lo = 0
    hi = (2 ** len(s)) - 1

    for c in s:
        if c == 'R': # upper
            lo = (lo + hi + 1) // 2
        else:
            hi = (lo + hi) // 2

    if s[-1] == 'R':
        return hi
    else:
        return lo

def get_row(s):
    lo = 0
    hi = (2 ** len(s)) - 1

    for c in s:
        if c == 'B': # upper
            lo = (lo + hi + 1) // 2
        else:
            hi = (lo + hi) // 2

    if s[-1] == 'B':
        return hi
    else:
        return lo

def get_id(line):
    row = get_row(line[:-3])
    col = get_col(line[-3:])
    return (row * 8) + col

def main():
    fname = '../data/p5.txt'

    with open(fname, 'r') as f:
        lines = f.readlines()
        max_id = 0
        max_line = None
        ids = []
        for line in lines:
            line = line.strip()
            sid = get_id(line)
            ids.append(sid)

            if sid > max_id:
                max_id = sid
                max_line = line

        print("Part one:", max_id)
        vals = sorted(ids)
        for idx, i in enumerate(vals):
            if idx > 0:
                if i - vals[idx - 1] > 1:
                    print("Part two:", i - 1)

if __name__ == '__main__':
    main()

File: julia/p5.jl
function get_row(line)
    lo = 0;
    hi = 2^length(line) - 1;
    for c in line
        if c == 'B'
            lo = div(lo + hi + 1, 2);
        else
            hi = div(lo + hi, 2);
        end
    end
    return lo;
end

function get_col(line)
    lo = 0;
    hi = 2^length(line) - 1;

    for c in line
        if c == 'R'
            lo = div(lo + hi + 1,  2);
        else
            hi = div(lo + hi, 2);
        end
    end
    return lo;
end

function get_id(line)
    row = get_row(line[begin:end-3]);
    col = get_col(line[end-2:end]);
    return 8row + col;
end

function main()
    fname = "../data/p5.txt";
    lines = readlines(fname);
    maxid = 0;
    ids = [];

    for l in lines
        id = get_id(l);
        maxid = max(id, maxid);
        push!(ids, id);
    end

    println("Part one: ", maxid);
    sort!(ids);
    for (idx, x) in enumerate(ids)
        if (idx > 1) && ((x - ids[idx - 1]) > 1)
            println("Part two: ", x - 1);
            break;
        end
    end
end

main();
